internals-js/src/operations.ts

```diff
@@ -1535,6 +1551,32 @@ export class SelectionSet {
     this._selections = mapValues(keyedSelections);
   }
 
+  minimizeSelectionSet(namedFragments: NamedFragments, fragmentSpreadsById: Map<string, FragmentSpreadSelection>): SelectionSet {
+    return this.lazyMap((selection) => {
+      if (selection.kind === 'FragmentSelection' && selection.element.typeCondition && selection.element.appliedDirectives.length === 0 && selection.selectionSet) {
+        const id = print(selection.toSelectionNode());
+        const existing = fragmentSpreadsById.get(id);
+        if (existing) return existing;
+
+        const minimizedSelectionSet = selection.selectionSet.minimizeSelectionSet(namedFragments, fragmentSpreadsById);
+        const fragmentDefinition = new NamedFragmentDefinition(
+          this.parentType.schema(),
+          `qp__${fragmentSpreadsById.size}`,
+          selection.element.typeCondition
+        ).setSelectionSet(minimizedSelectionSet);
+        const fragmentSpread = new FragmentSpreadSelection(this.parentType, namedFragments, fragmentDefinition, []);
```

Comment:
we cannot save `FragmentSpreadSelection` as it references its parent, instead we should cache `NamedFragmentDefinition` and recreate fragment spread selection for each reference

Reply:
[159d92c](https://github.com/apollographql/federation/pull/2958/commits/159d92c535a572415738d375f82a434b8e597a06)